src/Simulation/Simulators/OpenSystemsSimulator/DataModel/Extensions.cs

```diff
@@ -6,159 +6,217 @@
+                    // the kind of data for the object.
+                    // For variants of type unit, it's just the string.
+                    reader.Read();
+                    if (reader.TokenType == JsonTokenType.StartObject)
+                    {
+                        reader.Require(JsonTokenType.PropertyName);
+                        var kind = reader.GetString();
+                        // reader.GetString returns null when TokenType is
+                        // Null, but we've checked that condition above
+                        // by calling Require such that we can guarantee
+                        // kind is not null.
+                        Debug.Assert(kind is not null);
 
-                var propertyName = reader.GetString();
-
-                switch (propertyName)
-                {
-                    case "n_qubits":
-                        reader.Read();
-                        nQubits = reader.GetInt32();
-                        break;
-
-                    case "data":
-                        // In most cases, we expect an object with one property
-                        // indicating
-                        // the kind of data for the object.
-                        // For variants of type unit, it's just the string.
                         reader.Read();
-                        if (reader.TokenType == JsonTokenType.StartObject)
-                        {
-                            reader.Require(JsonTokenType.PropertyName);
-                            var kind = reader.GetString();
-
-                            reader.Read();
-                            completion = readData(ref reader, kind);
-
-                            // Finally, require an end to the object.
-                            reader.Require(JsonTokenType.EndObject);
-                        }
-                        else if (reader.TokenType == JsonTokenType.String)
-                        {
-                            var kind = reader.GetString();
-                            completion = readData(ref reader, kind);
-                        }
-                        else
-                        {
-                            throw new JsonException($"Expected either the start of an object or a string.");
-                        }
-                        break;
-
-                    default:
-                        throw new JsonException($"Unexpected property name {propertyName}.");
-                }
+
+                        completion = readData(ref reader, kind);
+
+                        // Finally, require an end to the object.
+                        reader.Require(JsonTokenType.EndObject);
+                    }
+                    else if (reader.TokenType == JsonTokenType.String)
+                    {
+                        var kind = reader.GetString();
+                        // The same logic follows here as well.
+                        Debug.Assert(kind is not null);
```

Comment:
(Just FYI)
I remember a few months ago Bettina complained that `Debug.Assert()` is undesirable for our test infra.
Later on I observed something like, if the `Debug.Assert()` is executed in the test and the assertion fails, then the test is terminated but it does not result in the CI run failure. The break stays unnoticed until 
someone actually runs the tests locally and observe the test failure, 
or looks at the CI run log.

Reply:
As the comment above notes, this assertion isn't actually used as an assertion per se, but to prove to the C# compiler that `kind` is non-null. That's not an invariant of `reader.GetString()`, such that the compiler believes that `kind` can be null at this point, but we have checked all conditions under which `GetString()` can be null such that we need to tell the compiler that to suppress warnings. This is identical to another case above in the same file, where a comment explains that in a bit more detail.